src/nvim/api/vimscript.c

```diff
@@ -745,3 +746,294 @@ Dictionary nvim_parse_expression(String expr, String flags, Boolean highlight, E
+  if (fp->uf_name[0] == K_SPECIAL) {
+    // Replace [ K_SPECIAL KS_EXTRA KE_SNR ] with "<SNR>"
+    name.size = name_len + 2;
+    name.data = xmalloc(name.size + 1);
+    name.data[0] = '<';
+    name.data[1] = 'S';
+    name.data[2] = 'N';
+    name.data[3] = 'R';
+    name.data[4] = '>';
+    memcpy(name.data + 5, fp->uf_name + 3, name_len - 3);
+    name.data[name.size] = '\0';
+  } else {
+    name.size = name_len;
+    name.data = xmalloc(name.size + 1);
+    memcpy(name.data, fp->uf_name, name_len);
+    name.data[name.size] = '\0';
+  }
+  return name;
+}
+
+/// Takes ownership of name.
+static Dictionary user_function_dict(ufunc_T *fp, String name, bool details, bool lines)
+{
+  Dictionary dict = ARRAY_DICT_INIT;
+  size_t dict_size = 2U + (details ? 8 : 0) + (lines ? 1 : 0);
+  kv_resize(dict, dict_size);
+
+  // Function name
+  PUT(dict, "name", STRING_OBJ(name));
+  PUT(dict, "type", CSTR_TO_OBJ("user"));
+
+  if (details) {
+    // Arguments
+    Array args = ARRAY_DICT_INIT;
+    if (fp->uf_args.ga_len > 0) {
+      kv_resize(args, (size_t)fp->uf_args.ga_len);
+      for (int j = 0; j < fp->uf_args.ga_len; j++) {
+        Dictionary arg = ARRAY_DICT_INIT;
+        PUT(arg, "name", CSTR_TO_OBJ((const char *)FUNCARG(fp, j)));
+        if (j >= fp->uf_args.ga_len - fp->uf_def_args.ga_len) {
+          PUT(arg, "default",
+              CSTR_TO_OBJ(((char **)(fp->uf_def_args.ga_data))
+                          [j - fp->uf_args.ga_len + fp->uf_def_args.ga_len]));
+        }
+        kv_push(args, DICTIONARY_OBJ(arg));
+      }
+    }
+    PUT(dict, "args", ARRAY_OBJ(args));
+    PUT(dict, "varargs", BOOLEAN_OBJ(fp->uf_varargs));
+    // Attributes
+    PUT(dict, "abort", BOOLEAN_OBJ(fp->uf_flags & FC_ABORT));
+    PUT(dict, "range", BOOLEAN_OBJ(fp->uf_flags & FC_RANGE));
+    PUT(dict, "dict", BOOLEAN_OBJ(fp->uf_flags & FC_DICT));
+    PUT(dict, "closure", BOOLEAN_OBJ(fp->uf_flags & FC_CLOSURE));
+    // Script
+    PUT(dict, "sid", INTEGER_OBJ(fp->uf_script_ctx.sc_sid));
+    PUT(dict, "lnum", INTEGER_OBJ(fp->uf_script_ctx.sc_lnum));
+  }
+
+  // Source lines
+  if (lines) {
+    Array array = ARRAY_DICT_INIT;
+    if (fp->uf_lines.ga_len > 0) {
+      kv_resize(array, (size_t)fp->uf_lines.ga_len);
+      for (int j = 0; j < fp->uf_lines.ga_len; j++) {
+        kv_push(array, CSTR_TO_OBJ((const char *)FUNCLINE(fp, j)));
+      }
+    }
+    PUT(dict, "lines", ARRAY_OBJ(array));
+  }
+
+  return dict;
+}
+
+/// Takes ownership of name.
+static Dictionary builtin_function_dict(const EvalFuncDef *fn, String name, bool details)
+{
+  Dictionary dict = ARRAY_DICT_INIT;
+  kv_resize(dict, details ? 7 : 2);
+
+  PUT(dict, "name", STRING_OBJ(name));
+  PUT(dict, "type", CSTR_TO_OBJ("builtin"));
+
+  if (details) {
+    PUT(dict, "min_argc", INTEGER_OBJ(fn->min_argc));
+    PUT(dict, "max_argc", INTEGER_OBJ(fn->max_argc));
+    PUT(dict, "base_arg", INTEGER_OBJ(fn->base_arg));
+    PUT(dict, "fast", BOOLEAN_OBJ(fn->fast));
```

Comment:
This info comes from `funcs.generated.h` generated by `gen_eval.lua`, which reads `eval.lua`.

For I think it's likely worth avoiding t What if we instead ,

1. move `eval.lua` to `runtime/lua/core/eval.lua`
2. we could make the field names a bit nicer like you've done here, e.g.
    - `assert_equal={args_min=2, args_max=3, base=2},`
    - instead of: `assert_equal={args={2, 3}, base=2},`
3. users can then can use both `require('vim.core.eval')` and `nvim_get_functions(builtin)`, and they return the same form.

## Benefits

- eliminates `builtin_function_dict()` ; becomes a Lua call in `nvim_get_functions()`.
- single source of truth: `eval.lua` is both the source and the final form. Helps with discovery.
- helps establish a pattern we've wanted for similar future work